AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -510,8 +663,24 @@ static OSStatus ExampleAVAudioEngineDeviceRecordCallback(void *refCon,
                              numFrames,
                              audioBufferList);
 
-    int8_t *audioBuffer = (int8_t *)audioBufferList->mBuffers[0].mData;
-    UInt32 audioBufferSizeInBytes = audioBufferList->mBuffers[0].mDataByteSize;
+    AudioBufferList *mixedAudioBufferList = context->mixedAudioBufferList;
+    assert(mixedAudioBufferList != NULL);
+    assert(mixedAudioBufferList->mNumberBuffers == audioBufferList->mNumberBuffers);
+    for(int i = 0; i < audioBufferList->mNumberBuffers; i++) {
+        mixedAudioBufferList->mBuffers[i].mNumberChannels = audioBufferList->mBuffers[i].mNumberChannels;
+        mixedAudioBufferList->mBuffers[i].mDataByteSize = audioBufferList->mBuffers[i].mDataByteSize;
+    }
+
+    OSStatus outputStatus = noErr;
+    AVAudioEngineManualRenderingBlock renderBlock = (__bridge AVAudioEngineManualRenderingBlock)(context->renderBlock);
+    const AVAudioEngineManualRenderingStatus ret = renderBlock(numFrames, mixedAudioBufferList, &outputStatus);
+
+    if (ret != AVAudioEngineManualRenderingStatusSuccess) {
+        NSLog(@"AVAudioEngine failed mix audio");
+    }
+
+    int8_t *audioBuffer = (int8_t *)mixedAudioBufferList->mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = mixedAudioBufferList->mBuffers[0].mDataByteSize;
```

Comment:
Should the AudioUnit know (from the context) whether or not it should be dealing with the mixing logic in the callback? At the moment it will attempt to do the mixing (via AVAudioEngine) if music is playing or not.

Reply:
As discussed, the change is recording does an extra copy when there is no audio file is being played.